src/impl_owned_array.rs

```diff
@@ -61,6 +61,55 @@ impl<A, D> Array<A, D>
 where
     D: Dimension,
 {
+    /// Returns the offset (in units of `A`) from the start of the raw data to
+    /// the first element, or `None` if the array is empty.
+    ///
+    /// In other words, after converting the array to `Vec<A>` with
+    /// [`.into_raw_vec()`](Self::into_raw_vec), this is the offset from the
+    /// start of the `Vec` to the first element of the array.
+    ///
+    /// ```
+    /// use ndarray::{Array2, Axis};
+    ///
+    /// let mut arr: Array2<f64> = Array2::zeros([3, 4]);
+    /// arr.slice_axis_inplace(Axis(0), (1..).into());
+    /// arr[[0, 0]] = 5.;
+    ///
+    /// let offset = arr.offset_to_first_elem().unwrap();
+    /// assert_eq!(offset, 4);
+    ///
+    /// let v = arr.into_raw_vec();
+    /// assert_eq!(v[offset], 5.);
+    /// ```
+    ///
+    /// In the case of zero-sized elements, the offset is somewhat meaningless.
+    /// For convenience, an offset will be returned such that all indices
+    /// computed using the offset, shape, and strides will be in-bounds for the
+    /// `Vec<A>` returned by [`.into_raw_vec()`](Self::into_raw_vec). Note that
+    /// this offset won't necessarily be the same as the offset for an array of
+    /// nonzero-sized elements sliced in the same way.
+    ///
+    /// ```
+    /// use ndarray::{Array2, Axis};
+    ///
+    /// let mut arr: Array2<()> = Array2::from_elem([3, 4], ());
+    /// arr.slice_axis_inplace(Axis(0), (1..).into());
+    /// let offset = arr.offset_to_first_elem().unwrap();
+    /// assert_eq!(offset, 0);
+    /// ```
+    pub fn offset_to_first_elem(&self) -> Option<usize> {
```

Comment:
Maybe we can use the name logic "offset from memory to first elem" again?

Reply:
Yeah, `offset_from_memory_to_ptr`, `offset_from_memory_to_first`, and `offset_from_memory_to_first_elem` are clearer than `offset_to_first_elem`. Another, perhaps better, option is to change `.into_raw_vec()` to return `(vec, offset)` instead of adding a new method. This would make it harder to misuse `.into_raw_vec()`.

Reply:
From the discussion in the other comment chain, I'm thinking that a good name would be `offset_from_alloc_to_logical_ptr`. How about making it an internal-only method for now and changing `.into_raw_vec()` to return `(vec, offset)` as I suggested above?

Reply:
You know, now that I think about it, I wonder if `std` will add a trait for `Vec`-like objects to minimize the proliferation of type parameters. The current `ndarray` API based on `ArrayBase` instead of traits can be inconvenient sometimes. Once custom (non-global) allocators are supported, `Vec` will be in a similar position since it'll be generic over the allocator.

Reply:
Makes sense.

I've long thought that we might parameterize the Array not just by the Vec's allocator but by wholly new kinds of owned arrays - let's say some other allocation system, like always 64-byte aligned or what it could be, or maybe always using some external specific allocation function. In those cases we don't use Vec or std allocation traits at all, maybe.

Reply:
Yeah, an `ArrayVec`-backed n-D array type would be useful too. Even with additional owned types/parameters, though, we could provide `.into_raw_vec()` for only the type parameters that make sense. Long-term, I'd like to move away from `ArrayBase` to a set of traits describing n-D array functionality, implemented for the various array types.

Reply:
Fully agreed. It just seems to be a lot of work getting to the traitification. Maybe we can start it small and do it piece by piece, somehow?

Reply:
Yeah, I think we can do it piece-by-piece. We can introduce traits one-by-one and implement them for `ArrayBase`. For example, we can start with an unsafe trait to get the ptr, dim, and strides, and then add a trait using it for readonly indexing/slicing (with methods returning element references and `ArrayView`s). I expect it'll take multiple iterations to fully refine the set of traits we need.